refactor(napi/oxlint): diagnostics communicate which rule via rule index, not rule ID

overlookmotel opened the pull request:
Related to #12437.

Diagnostics sent from JS to Rust specify which rule the diagnostic relates to via an index into `external_rules`, instead of the `ExternalRuleId`. This allows removing the loop over `external_rules` on Rust side to find the relevant rule.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12482?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#12482** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12482?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12482?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#12481** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12481?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12480** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12480?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#12479** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12479?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12482).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camc314 commented on crates/oxc_linter/src/external_plugin_store.rs:
```diff
@@ -79,8 +79,11 @@ impl ExternalPluginStore {
 
-    pub fn resolve_plugin_rule_names(&self, external_rule_id: u32) -> Option<(&str, &str)> {
-        let external_rule = self.rules.get(ExternalRuleId::from_raw(external_rule_id))?;
+    pub fn resolve_plugin_rule_names(
+        &self,
+        external_rule_id: ExternalRuleId,
+    ) -> Option<(&str, &str)> {
+        let external_rule = self.rules.get(external_rule_id)?;
```
if the argument here is `ExternalRuleId`, I don't think we need to use `.get` since we're indexing into the index vec with the correct key

camc314 replied:
the reason this returned an option originally is because you could pass in an arbitary `u32` value which might not be in the correct range.

(i guess you could pass in an arbitary value for `ExternalRuleId`, but it's a lot less likley

overlookmotel replied:
Good point. While there's no real difference between `u32` and `ExternalRuleId`, the ID now comes from within Rust code, rather than being passed from JS, so we can have more confidence that's it's valid.

I've pushed a commit to make `resolve_plugin_rule_names` infallible.

camc314 commented on napi/oxlint2/src-js/index.js:
```diff
@@ -51,41 +51,39 @@ async function loadPluginImpl(path) {
     ruleNames.push(ruleName);
-    registeredRules.push({ rule, context: new Context(ruleId, `${pluginName}/${ruleName}`) });
-    ruleId++;
+    registeredRules.push({ rule, context: new Context(`${pluginName}/${ruleName}`) });
```
unrelated, but just noticed, should we re-use the string for the `pluginName/ruleName` rather than creating it? might need benching

overlookmotel replied:
How do you mean? This is when rule is registered, which only happens once. So I think each string is only created once.

We could do this:

```js
const pluginNameWithSlash = `${pluginName}/`;
for (const [ruleName, rule] of Object.entries(plugin.rules)) {
  ruleNames.push(ruleName);
  registeredRules.push({ rule, context: new Context(`${pluginNameWithSlash}${ruleName}`) });
}
```

But I'm not sure if that'd be a perf improvement or not. If it is, I doubt the difference would be measurable.

camc314 replied:
probably won't have a perf diff, but i more meant changing `plugin.rules shape to be like
```
type Rules = { [ruleName: string]: { fullName: `${pluginName}/${ruleNme}`, rule: any } }
```

so we only construct `fullName` once. not important to optimize right now though.

Currently we construct `${pluginNameWithSlash}${ruleName}`` num files * num rules times, but we could just change it to construct a max of `num rules` times.

camc314 commented on napi/oxlint2/src-js/index.js:
```diff
@@ -123,6 +121,14 @@ class Context {
     return this.physicalFilename;
   }
+
+  static {
+    setupContextForFile = (context, ruleIndex, filePath) => {
+      context.#ruleIndex = ruleIndex;
+      context.filename = filePath;
+      context.physicalFilename = filePath;
+    };
+  }
```
out of interest is static faster than a non-static method directly on the class??

camc314 replied:
lgtm

overlookmotel committed:
refactor(napi/oxlint): diagnostics communicate which rule via rule index, not rule ID (#12482)

Related to #12437.

Diagnostics sent from JS to Rust specify which rule the diagnostic relates to via an index into `external_rules`, instead of the `ExternalRuleId`. This allows removing the loop over `external_rules` on Rust side to find the relevant rule.
